### PR TITLE
CODENVY-40: add pagination to list of users

### PR DESCRIPTION
**Pull request by olexii4**
Signed-off-by: Oleksii Orel <oorel@codenvy.com>

### CODE REVIEW

File: dashboard/src/components/api/codenvy-user.factory.js [hunk 1]
```diff
@@ -43,7 +45,24 @@ export class CodenvyUser {
...
-      getUsers: {method: 'GET', url: '/api/admin/user?maxItems=:maxItems&skipCount=:skipCount', isArray: true},
+      getUsers: {
+        method: 'GET',
+        url: '/api/admin/user?maxItems=:maxItems&skipCount=:skipCount',
+        isArray: false,
+        responseType: 'json',
+        transformResponse: (data , headersGetter) => {
+          let links = new Map();
+          let headersLink;
+          if(headersLink = headersGetter('link')){
```
Comment (benoitf) [inline dashboard/src/components/api/codenvy-user.factory.js]: missing space around parenthesis

File: dashboard/src/components/api/codenvy-user.factory.js [hunk 2]
```diff
@@ -43,7 +45,24 @@ export class CodenvyUser {
...
+            let pattern = new RegExp('<([^>]+?)>.+?rel="([^"]+?)"', 'g');
+            let result;
+            while (result = pattern.exec(headersLink)) { //look for pattern
```
Comment (benoitf) [inline dashboard/src/components/api/codenvy-user.factory.js]: let should be inside the while (let result = ) no ?

File: dashboard/src/components/api/codenvy-user.factory.js [hunk 3]
```diff
@@ -43,7 +45,24 @@ export class CodenvyUser {
...
+      getUsers: {
+        method: 'GET',
+        url: '/api/admin/user?maxItems=:maxItems&skipCount=:skipCount',
+        isArray: false,
+        responseType: 'json',
+        transformResponse: (data , headersGetter) => {
+          let links = new Map();
+          let headersLink;
+          if(headersLink = headersGetter('link')){
+            let pattern = new RegExp('<([^>]+?)>.+?rel="([^"]+?)"', 'g');
```
Comment (benoitf) [inline dashboard/src/components/api/codenvy-user.factory.js]: we should need a unit test for this regexp

File: dashboard/src/components/api/codenvy-user.factory.js [hunk 4]
```diff
@@ -86,13 +111,111 @@ export class CodenvyUser {
...
+      this.usersMap.set(user.id, user);//add user
       this.imsLicenseApi.fetchLicenseLegality();//fetch license legality
     });
 
     return promise;
   }
 
+  _getPageParamByLink(pageLink) {
+    let lastPageParamMap = new Map();
+    let pattern = new RegExp('([_\\w]+)=([\\w]+)', 'g');
```
Comment (benoitf) [inline dashboard/src/components/api/codenvy-user.factory.js]: we should need a unit test for this regexp

File: dashboard/src/components/api/codenvy-user.factory.js [hunk 5]
```diff
@@ -86,13 +111,111 @@ export class CodenvyUser {
...
+    if(!data.links){
+      return;
+    }
+    let firstPageLink = data.links.get('first');
+    if (firstPageLink) {
+      let firstPageData = {link: firstPageLink};
+      if (this.pageInfo.currentPageNumber === 1) {
+        firstPageData.users = data.users;
+      }
+      if(!this.userPagesMap.get(1) || firstPageData.users){
```
Comment (benoitf) [inline dashboard/src/components/api/codenvy-user.factory.js]: spaces

File: dashboard/src/components/api/codenvy-user.factory.js [hunk 6]
```diff
@@ -86,13 +111,111 @@ export class CodenvyUser {
...
+        this.userPagesMap.set(1, firstPageData);
+      }
+    }
+    let lastPageLink = data.links.get('last');
+    if (lastPageLink) {
+      let pageParam = this._getPageParamByLink(lastPageLink);
+      this.pageInfo.countOfPages = pageParam.skipCount/pageParam.maxItems + 1;
```
Comment (benoitf) [inline dashboard/src/components/api/codenvy-user.factory.js]: spaces

File: dashboard/src/components/api/codenvy-user.factory.js [hunk 7]
```diff
@@ -86,13 +111,111 @@ export class CodenvyUser {
...
+      let lastPageData = {link: lastPageLink};
+      if (this.pageInfo.currentPageNumber === this.pageInfo.countOfPages) {
+        lastPageData.users = data.users;
+      }
+      if(!this.userPagesMap.get(this.pageInfo.countOfPages) || lastPageData.users){
```
Comment (benoitf) [inline dashboard/src/components/api/codenvy-user.factory.js]: spaces

File: dashboard/src/components/api/codenvy-user.factory.js [hunk 8]
```diff
@@ -86,13 +111,111 @@ export class CodenvyUser {
...
+        this.userPagesMap.set(this.pageInfo.countOfPages, lastPageData);
+      }
+    }
+    let prevPageLink = data.links.get('prev');
+    let prevPageNumber = this.pageInfo.currentPageNumber - 1;
+    if (prevPageNumber > 0 && prevPageLink) {
+      let prevPageData = {link: prevPageLink};
+      if(!this.userPagesMap.get(prevPageNumber)){
```
Comment (benoitf) [inline dashboard/src/components/api/codenvy-user.factory.js]: spaces

File: dashboard/src/components/api/codenvy-user.factory.js [hunk 9]
```diff
@@ -86,13 +111,111 @@ export class CodenvyUser {
...
+        this.userPagesMap.set(prevPageNumber, prevPageData);
+      }
+    }
+    let nextPageLink = data.links.get('next');
+    let nextPageNumber = this.pageInfo.currentPageNumber + 1;
+    if (nextPageNumber) {
+      let lastPageData = {link: nextPageLink};
+      if(!this.userPagesMap.get(nextPageNumber)){
```
Comment (benoitf) [inline dashboard/src/components/api/codenvy-user.factory.js]: spaces

File: dashboard/src/components/api/codenvy-user.factory.js [hunk 10]
```diff
@@ -103,22 +226,72 @@ export class CodenvyUser {
...
+      pageNumber = this.pageInfo.countOfPages;
+    }
+    let pageData = this.userPagesMap.get(pageNumber);
+    if (pageData.link) {
+      this.pageInfo.currentPageNumber = pageNumber;
+      let promise;
+      if (this.codenvyDashboardConfig.developmentMode) {
+        promise = this.remoteUserAPI.getUsers(this._getPageParamByLink(pageData.link)).$promise;
+      } else {
+        promise = this.$http.get(pageData.link).$promise;
```
Comment (benoitf) [inline dashboard/src/components/api/codenvy-user.factory.js]: so we need $http ?

File: dashboard/src/components/api/codenvy-user.factory.js (hunk 2 above)
Comment (olexii4) [inline dashboard/src/components/api/codenvy-user.factory.js]: Not for "while"

File: dashboard/src/components/api/codenvy-user.factory.js (hunk 3 above)
Comment (olexii4) [inline dashboard/src/components/api/codenvy-user.factory.js]: done

File: dashboard/src/components/api/codenvy-user.factory.js (hunk 4 above)
Comment (olexii4) [inline dashboard/src/components/api/codenvy-user.factory.js]: done

File: dashboard/src/components/api/codenvy-user.factory.js (hunk 10 above)
Comment (olexii4) [inline dashboard/src/components/api/codenvy-user.factory.js]: Ok. I have removed $http

File: dashboard/src/components/api/codenvy-user.factory.js (hunk 9 above)
Comment (olexii4) [inline dashboard/src/components/api/codenvy-user.factory.js]: ok

File: dashboard/src/components/api/codenvy-user.factory.js (hunk 8 above)
Comment (olexii4) [inline dashboard/src/components/api/codenvy-user.factory.js]: ok

File: dashboard/src/components/api/codenvy-user.factory.js (hunk 7 above)
Comment (olexii4) [inline dashboard/src/components/api/codenvy-user.factory.js]: ok

File: dashboard/src/components/api/codenvy-user.factory.js (hunk 6 above)
Comment (olexii4) [inline dashboard/src/components/api/codenvy-user.factory.js]: ok

File: dashboard/src/components/api/codenvy-user.factory.js (hunk 5 above)
Comment (olexii4) [inline dashboard/src/components/api/codenvy-user.factory.js]: ok

File: dashboard/src/components/api/codenvy-user.factory.js (hunk 1 above)
Comment (olexii4) [inline dashboard/src/components/api/codenvy-user.factory.js]: ok

**Comment by benoitf** [comment]
OK now thx

File: dashboard/src/app/admin/user-management/user-management.controller.js [hunk 11]
```diff
@@ -44,15 +44,19 @@ export class AdminsUserManagementCtrl {
...
-    });
-  }
+      this.isLoading = true;
+      codenvyAPI.getUser().fetchUsers(this.maxItems, this.skipCount).then(() => {
+        this.isLoading = false;
+        this.updateUsers();
+      }, (error) => {
+        this.isLoading = false;
+        if (error && error.status !== 304) {
+          this.cheNotification.showError(error.data && error.data.message ? error.data.message : 'Error.');
```
Comment (ashumilova) [inline dashboard/src/app/admin/user-management/user-management.controller.js]: "Error." - Failed to load users or Failed to retrieve the list of users

File: dashboard/src/app/admin/user-management/user-management.controller.js [hunk 12]
```diff
@@ -67,23 +71,48 @@ export class AdminsUserManagementCtrl {
...
+        this.cheNotification.showError(error.data && error.data.message ? error.data.message : 'Update information failed.');
       }
     });
   }
 
   /**
+   * Returns true if the next page is exist.
+   * @returns {boolean}
+   */
+  noNextPage() {
```
Comment (ashumilova) [inline dashboard/src/app/admin/user-management/user-management.controller.js]: noNextPage -> hasNextPage , then "Returns true if the next page exists."

File: dashboard/src/app/admin/user-management/user-management.controller.js [hunk 13]
```diff
@@ -67,23 +71,48 @@ export class AdminsUserManagementCtrl {
...
+    return this.pagesInfo.currentPageNumber === this.pagesInfo.countOfPages;
+  }
+
+  /**
+   * Returns true if the previous page is exist.
+   * @returns {boolean}
+   */
+  noPreviousPage() {
```
Comment (ashumilova) [inline dashboard/src/app/admin/user-management/user-management.controller.js]: noPreviousPage -> hasPreviousPage, then "Returns true if the previous page exists."

**Pull request merged**